### ISSUE TITLE
Autocomplete/Intellisense for Keys

### ISSUE THREAD

**Issue by TheSentry** [SeriousM/WPFLocalizationExtension#99]
Hi!

This is my first time developing a C# app in Visual Studio and I want to do it right, so I'm localizing it. Problem is, autocomplete (intellisense) won't work for the WPF Localization Extension.

For example, when I have this code

        <MenuItem Header="{lex:Loc MenuFile}">
and my key/value pairs in my Resources.resx file, instead of "MenuFile", it only suggests
* Converter
* ConverterParameter
* ForceCulture
* Key

after `lex:Loc` If I use the "Key=" Syntax, I get nothing.

        <MenuItem Header="{lex:Loc Key=MenuFile}">
Is there any way to fix this? Or is this a major feature to implement? The code itself works, btw.

Thx!

**Comment by SeriousM** [comment]
Hi! 

Visual studio or blend does not know what to offer because it does not know how the extension actually works. 
Intellisense was one of my points to implement, but since we have the ability to use different localization sources (csv, database, etc) it's kinda impossible to offer keys from an dynamic source. In addition to this, the data source is usually known during the execution in designtime and runtime, and the Intellisense is using only the written code, not created assemblies. 

To sum it up: its sadly impossible.

**Comment by TheSentry** [comment]
> To sum it up: its sadly impossible. 

That's a shame. But how impossible is impossible? Is it really impossible?

As I said, I'm new to C# and Visual Studio, but I've found this [stackoverflow answer](http://stackoverflow.com/a/10527253) that says it might be possible. I could give it a try, but I wouldn't get my hope up if I where you (or anyone else).

Thanks for the very fast response, btw!

**Comment by UeliHilfiker** [comment]
I'm using the **[Automatic key retrieval](Automatic key retrieval)** - feature, so autocomplete is not needed. ;-)

**Comment by SeriousM** [comment]
> That's a shame. But how impossible is impossible? Is it really impossible?
> 
> As I said, I'm new to C# and Visual Studio, but I've found this stackoverflow answer that says it might be possible. I could give it a try, but I wouldn't get my hope up if I where you (or anyone else).

Well, yes it is.
There are several problems:
- the source of the available keys is dynamic (defined on the binding and/or on the parent element up to the window during runtime and/or by code on startup)
- there is no way to cache loaded keys in case of a hard-to-call resource (eg. database, webservice)
- the loading of the localize extension dll is non-deterministic because VS as well as Blend don't give a shit about third-party extensions. Therefore we have already problems to get all the updates done on the documents

Sorry.



> I'm using the Automatic key retrieval - feature, so autocomplete is not needed. ;-)

Glad to hear that someone uses this feature! (jk jk)

**Comment by TheSentry** [comment]
Ok, thank you all for the input. It helped me to get a better picture of the situation.

> I'm using the Automatic key retrieval - feature, so autocomplete is not needed. ;-)

Well, this just turns the problem on its head, because now I'd like autocomplete to make sure the key I write in the resx file is the same as I named the element.
